Support stacked branches

stefanhaller opened the pull request:
- **PR Description**

If the user has `rebase.updateRefs` in their git config, working with stacked branches behaves as expected now. For example, pressing "e" or "d" in a lower branch now preserves the stack.

Fixes most of #2527.

- **Please check if the PR fulfills these requirements**

* [x] Cheatsheets are up-to-date (run `go run scripts/cheatsheet/main.go generate`)
* [x] Code has been formatted (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#code-formatting))
* [x] Tests have been added/updated (see [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/integration/README.md) for the integration test guide)
* [x] Text is internationalised (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#internationalisation))
* [x] Docs (specifically `docs/Config.md`) have been updated if necessary
* [x] You've read through your own file changes for silly mistakes etc

stefanhaller commented:
Keeping as a draft for now as long as it is sitting on https://github.com/jesseduffield/lazygit/pull/2370, and waiting for an update to git-todo-parser. Apart from these, I consider it ready though.

jesseduffield commented on pkg/app/daemon/daemon.go:
```diff
@@ -34,6 +37,15 @@ const (
+	// The existing action of the todo to be changed is expected to be "pick".
+	//
+	// If this is used, the value of RebaseTODOEnvKey must be empty.
+	ChangeTodoActionEnvKey string = "LAZYGIT_CHANGE_TODO_ACTION"
```
We've got a few env vars now, and it's not ideal working with an interface that consists of multiple env vars. What are your thoughts on having a struct that contains all the args we need and JSONifying it into a single env var then converting back to a struct here. That allows for arbitrary complexity in the interface

stefanhaller replied:
Yes, good idea. I squashed this into the last commit (see ea2583ef590), using fixups didn't seem to make much sense for this.

jesseduffield commented on pkg/app/daemon/daemon.go:
```diff
@@ -94,19 +106,52 @@ func (self *rebaseDaemon) Run() error {
+}
+
+func (self *rebaseDaemon) actionFromString(actionString string) todo.TodoCommand {
+	for t := todo.Pick; t < todo.Comment; t++ {
```
Worth having a comment saying that we're iterating through all possible todo kinds (it's not obvious that Pick == 0 and Comment is the last one).

In fact, it would be good if the package itself could export a slice of all the different values so that we can iterate through it without needing to worry about it getting out of sync. But obvs not required for this PR

stefanhaller replied:
This is obsolete, `actionFromString` is no longer needed after the json refactor.

jesseduffield commented on pkg/app/daemon/daemon.go:
```diff
@@ -94,19 +106,52 @@ func (self *rebaseDaemon) Run() error {
 	}
 
-	return os.WriteFile(path, todoContent, 0o644)
+	return 0
```
Rather than use 0 as a sentinel value, I'd have the function return `(todo.TodoCommand, bool)` where a false value means we couldn't find a match. That defends us against the situation where 0 becomes a valid value in that enum. In general I don't like depending on the underlying values of a 3rd party enum cos it's subject to change and it's hard to reason about without going and looking at the enum definition.

While I'm on the topic, I wonder if we should be mapping from the todo package's todo type to our own in a central location to reduce our exposure to this stuff. I don't know how much effort that would be, I'm happy to be told it's overkill.

stefanhaller replied:
Also obsolete now.

As for the question of creating our own enum and mapping it to the todo package's todo type, I did consider this when I turned `Commit.Action` into an enum. We would have been forced to do that if `todo.TodoCommand` wasn't one-based. I decided that it's not necessary because it's so unlikely that the todo package will suddenly change its enum to start at 0. And it can be even more error-prone, because the mapping would have to be maintained manually, so if the todo package's enum changes (as it recently did in https://github.com/fsmiamoto/git-todo-parser/pull/2), we might not notice and end up with bogus mappings.

jesseduffield commented on pkg/commands/git_commands/rebase.go:
```diff
@@ -136,11 +141,17 @@ func (self *RebaseCommands) EditRebase(branchRef string) error {
+	todoLines        []TodoLine
+	overrideEditor   bool
+	prepend          bool
+	changeTodoAction []ChangeTodoAction
```
Can we rename this to `changeTodoActions` or `todoActionChanges`?

stefanhaller replied:
Yes, even though it goes away again at the end of the branch. Still, I like clean history, so I fixed it in the middle. Had to squash it in though, fixups wouldn't work for this.

jesseduffield commented on pkg/app/daemon/daemon.go:
```diff
@@ -46,13 +47,24 @@ const (
+var logFile io.StringWriter
+
 func Handle(common *common.Common) {
+	logFile, _ = os.Create("/tmp/daemon-log.txt")
```
Could you explain the rationale behind this?

jesseduffield reviewed (changes requested):
Looking good, I left some comments

stefanhaller commented on pkg/commands/git_commands/patch.go:
```diff
@@ -107,9 +108,11 @@ func (self *PatchCommands) MovePatchToSelectedCommit(commits []*models.Commit, s
-			{sha: commits[sourceCommitIdx].Sha, newAction: todo.Edit},
-			{sha: commits[destinationCommitIdx].Sha, newAction: todo.Edit},
+		instruction: ChangeTodoActionsInstruction{
+			actions: []daemon.ChangeTodoAction{
```
It's a little awkward that `daemon` now appears in client code like this, but I had to move the struct to `daemon.go` for the json unmarshalling. Any better ideas how to do this? I suppose I could move the struct to `utils/RebaseTodo.go` instead, not sure that's any better though.

jesseduffield commented:
@stefanhaller I've played around with this a bit locally and I've pushed a commit which changes the interface a bit: now there's one daemon kind for each action you might want to perform and the construction, (de)serialisation and running of instructions is all in one place (the daemon package), which I think makes sense. All outside packages need to know is what to pass when creating an instruction for the daemon, and to add the env vars to the command.

One downside of this approach is that we're no longer coupling the logging with the instruction (the daemon package doesn't care what you log) and I think that's okay.

Let me know your thoughts.

stefanhaller reviewed (commented):
Nice, that's a good improvement, thanks. You'll probably want to squash this with the previous refactor commit so that the history is a little less confusing.

I'll force-push now to bump the git-todo-parser version in the first commit. Ready to go from my side after that.

stefanhaller commented:
I couldn't help adding another commit here to fix #2414, now that it's so easy to extend the daemon with new functionality. Really great work on that last refactoring! See bdd7e6a4ca0f.

jesseduffield commented:
Sorry it's taken me a while to get back to this. I've pushed a refactor of moveFixupCommitDown, and added some unit tests to ensure it has the same behaviour as your implementation. I find it easier to read when the validation logic all happens up-front rather than being mixed in with the non-validation logic. Lemme know your thoughts

stefanhaller commented:
Makes sense, thanks for the improvement. I added a fixup to rename `moveFixupCommitDownAux` to just `moveFixupCommitDown`, as that's the naming convention we already use for the other functions in this file (e.g. `MoveTodoDown`/`moveTodoDown`); also, it moves your new tests to the existing test file (`rebaseTodo_test.go`) and delete the new test file again (`rebase_todo_test.go`). See 29b2d31f507c

Added another fixup to address [this comment](https://github.com/jesseduffield/lazygit/pull/2552#discussion_r1167740741) above: 032a48d2310b

stefanhaller commented:
I squashed the fixups right away, hoping they are uncontroversial, and rebased on master. Ready to go from my side.

stefanhaller commented:
I added a new commit at the beginning of the branch that should fix the failing test: 8f1e26c187c4. @Ryooooooga, can you please have a look, it's related to the revert of the `--force-if-includes` PR.

jesseduffield commented:
@stefanhaller I'm gonna go ahead and merge this. If @Ryooooooga takes issue with your fix we can address that in a separate PR
